crates/ra_ide/src/syntax_highlighting.rs

```diff
@@ -555,3 +620,130 @@ fn highlight_injection(
 
     Some(())
 }
+
+fn highlight_doctest_element(stack: &mut HighlightedRangeStack, element: SyntaxElement) {
+    // Iterate over all ast nodes that implement `DocCommentsOwner`
+    if let Some(node) = element.as_node().cloned().and_then(ast::SourceFile::cast) {
+        highlight_doctests(stack, node);
+    } else if let Some(node) = element.as_node().cloned().and_then(ast::FnDef::cast) {
+        highlight_doctests(stack, node);
+    } else if let Some(node) = element.as_node().cloned().and_then(ast::StructDef::cast) {
+        highlight_doctests(stack, node);
+    } else if let Some(node) = element.as_node().cloned().and_then(ast::UnionDef::cast) {
+        highlight_doctests(stack, node);
+    } else if let Some(node) = element.as_node().cloned().and_then(ast::RecordFieldDef::cast) {
+        highlight_doctests(stack, node);
+    } else if let Some(node) = element.as_node().cloned().and_then(ast::EnumDef::cast) {
+        highlight_doctests(stack, node);
+    } else if let Some(node) = element.as_node().cloned().and_then(ast::EnumVariant::cast) {
+        highlight_doctests(stack, node);
+    } else if let Some(node) = element.as_node().cloned().and_then(ast::TraitDef::cast) {
+        highlight_doctests(stack, node);
+    } else if let Some(node) = element.as_node().cloned().and_then(ast::Module::cast) {
+        highlight_doctests(stack, node);
+    } else if let Some(node) = element.as_node().cloned().and_then(ast::ConstDef::cast) {
+        highlight_doctests(stack, node);
+    } else if let Some(node) = element.as_node().cloned().and_then(ast::StaticDef::cast) {
+        highlight_doctests(stack, node);
+    } else if let Some(node) = element.as_node().cloned().and_then(ast::TypeAliasDef::cast) {
+        highlight_doctests(stack, node);
+    } else if let Some(node) = element.as_node().cloned().and_then(ast::ImplDef::cast) {
+        highlight_doctests(stack, node);
+    } else if let Some(node) = element.as_node().cloned().and_then(ast::MacroCall::cast) {
+        highlight_doctests(stack, node);
+    }
+}
+
+fn highlight_doctests<N: DocCommentsOwner>(stack: &mut HighlightedRangeStack, node: N) {
```

Comment:
I would expect this function as a composition of the two:

```rust
fn extract_doc_comments(node: impl DocCOmmentsOwner) -> Option<Vec<(String, RangesMap)>> {}
fn highlight_doc_comment(text: String, rangesMap: RangesMap, stack: &mut HighlightedRangeStack) {}
```

It seems that, by extracting the pure functioncal part of collecting the text of examples and computing a range mapping, we should make this more testable&understandable?